File: structures/trie.cs
using System;
using System.Collections.Generic;

namespace geeks
{
    public class Trie
    {
        private List<int[]> tree;
        private List<bool> end;
        private List<int> count;
        private int nodes;

        public Trie()
        {
            nodes = 1;
            tree = new List<int[]>();
            tree.Add(new int[52]);
            for (int i = 0; i < 52; i++)
                tree[0][i] = -1;

            end = new List<bool> { false };
            count = new List<int> { 0 };
        }

        private int GetIndex(char c)
        {
            if (c >= 'a' && c <= 'z')
                return c - 'a';
            else if (c >= 'A' && c <= 'Z')
                return c - 'A' + 26;
            else
                return -1;
        }

        public void Insert(string word)
        {
            int current = 0;

            foreach (char c in word)
            {
                int index = GetIndex(c);

                if (index < 0 || index >= 52)
                    continue;

                if (tree[current][index] == -1)
                {
                    tree[current][index] = nodes;

                    int[] newNode = new int[52];
                    for (int i = 0; i < 52; i++)
                        newNode[i] = -1;

                    tree.Add(newNode);
                    end.Add(false);
                    count.Add(0);
                    nodes++;
                }

                current = tree[current][index];
                count[current]++;
            }

            end[current] = true;
        }

        public bool Search(string word)
        {
            int current = 0;

            foreach (char c in word)
            {
                int index = GetIndex(c);

                if (index < 0 || index >= 52 || tree[current][index] == -1)
                {
                    return false;
                }

                current = tree[current][index];
            }

            return end[current];
        }

        public bool StartsWith(string prefix)
        {
            int current = 0;

            foreach (char c in prefix)
            {
                int index = GetIndex(c);

                if (index < 0 || index >= 52 || tree[current][index] == -1)
                {
                    return false;
                }

                current = tree[current][index];
            }

            return true;
        }

        public int CountWords(string prefix)
        {
            int current = 0;

            foreach (char c in prefix)
            {
                int index = GetIndex(c);

                if (index < 0 || index >= 52 || tree[current][index] == -1)
                {
                    return 0;
                }

                current = tree[current][index];
            }

            return count[current];
        }
    }

}

File: structures/trie.cpp
#include <bits/stdc++.h>

using namespace std;

class trie
{
public:
    int nodes;
    vector<vector<int>> tree;
    vector<bool> end;
    vector<int> count;
    vector<string> words;

    trie()
    {
        nodes = 1;
        tree.push_back(vector<int>(26, -1));
        end.push_back(false);
        count.push_back(0);
    }

    void insert(const string &s)
    {
        int node = 0;

        for (char c : s)
        {
            int index = c - 'a';

            if (tree[node][index] == -1)
            {
                tree[node][index] = nodes++;
                tree.push_back(vector<int>(26, -1));
                end.push_back(false);
                count.push_back(0);
            }

            node = tree[node][index];
            count[node]++;
        }

        end[node] = true;
        words.push_back(s);
    }

    bool searh(const string &s)
    {
        int node = 0;

        for (char c : s)
        {
            int index = c - 'a';

            if (tree[node][index] == -1)
            {
                return false;
            }

            node = tree[node][index];
        }

        return end[node];
    }

    bool starts_with(const string &prefix)
    {
        int node = 0;

        for (char c : prefix)
        {
            int index = c - 'a';

            if (tree[node][index] == -1)
            {
                return false;
            }

            node = tree[node][index];
        }

        return true;
    }

    bool prefix_query(const string &prefix)
    {
        int node = 0;

        for (char c : prefix)
        {
            int index = c - 'a';

            if (tree[node][index] == -1)
            {
                return 0;
            }

            node = tree[node][index];
        }

        return count[node];
    }

    vector<string> lexicographic_order()
    {
        words.clear();
        dfs(0, "");
        return words;
    }

    void dfs(int node, string current)
    {
        if (end[node])
        {
            words.push_back(current);
        }

        for (int i = 0; i < 26; i++)
        {
            if (tree[node][i] != -1)
            {
                char c = i + 'a';

                dfs(tree[node][i], current + c);
            }
        }
    }
};